### PR TITLE
Calendar

### PR DESCRIPTION
**Pull request by ESaracay**
Week 5 Algorithm

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -15,10 +15,157 @@
...
+
+    // Returns timeranges with optional attendess if any exist 
+    return (optionalTimes.size() == 0) ? availableTimes : optionalTimes;
   }
-}
+  
+  /*
+  * Sorts events by start time and filters out events that mandatoryAttendees are not apart of.
+  */
+  public ArrayList<Event> mySort(ArrayList<Event> list, HashSet<String> mandatoryAttendees) {
```
Comment (zmr) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: consider using a more descriptive, impersonal name (e.g. eventSort, for example)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -15,10 +15,157 @@
...
+      if (previousSlot == null || !eventTimeRange.overlaps(previousSlot)) {
+        availableTimes.add(TimeRange.fromStartEnd(rangeStart, eventTimeRange.start(), false));
+        rangeStart = eventTimeRange.end();
+        previousSlot = eventTimeRange;
+      } else if(previousSlot != null && previousSlot.contains(eventTimeRange)) {
+        // Checks for nested events
+        rangeStart = previousSlot.end();
+      } else {
+        rangeStart = eventTimeRange.end();
+        previousSlot = eventTimeRange;
```
Comment (zmr) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: looks like lines 46-47 are the same as 40-41. You should be able to condense these.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -15,10 +15,157 @@
...
+      for (TimeRange range: optionalSlots.keySet()) {
+        if (optionalSlots.get(range) == maxNumOpt) {
+          maxOptionalTimeRanges.add(range);
+        }
+      }
+
+    return maxOptionalTimeRanges;     
+    }
+
+    /*
```
Comment (tylerhou) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Wrong JavaDoc style.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,10 +15,157 @@
...
+
+      // Adds timeRanges of the observed max number of optionals to the ArrayList
+      ArrayList<TimeRange> maxOptionalTimeRanges = new ArrayList<>();
+      for (TimeRange range: optionalSlots.keySet()) {
+        if (optionalSlots.get(range) == maxNumOpt) {
+          maxOptionalTimeRanges.add(range);
+        }
+      }
+
+    return maxOptionalTimeRanges;
```
Comment (tylerhou) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Indentation.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -15,10 +15,157 @@
...
+  public ArrayList<Event> mySort(ArrayList<Event> list, HashSet<String> mandatoryAttendees) {
+    ArrayList<Event> newList = new ArrayList<>();
+    while (list.size() > 0) {
+      int indexMin = 0;
+      for (int i = 1; i < list.size(); i++) {
+        int newTime = list.get(i).getWhen().start(); 
+        if (list.get(indexMin).getWhen().start() > newTime) {
+          indexMin = i;
+        }
+      }
```
Comment (tylerhou) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Seems pretty n^2 — can you sort then filter by attendees? That should get you to n log n.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -15,10 +15,157 @@
...
+      // Cycles through times that work for mandatory attendees
+      int maxNumOpt = 0;
+      // HashMap contains timetange's and their corresponding number of optional attendees available
+      HashMap<TimeRange, Integer> optionalSlots = new HashMap<TimeRange, Integer>();
+      for (TimeRange freeTimeRange: mandatoryAttendeesTimes) {
+        int bucketStart = freeTimeRange.start();
+        int endOfSlot = freeTimeRange.end() - 1;
+        int previousNumOpt = 0;
+        System.out.println(freeTimeRange);
```
Comment (tylerhou) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Remove prints.

**Review by tylerhou — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (ESaracay) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (ESaracay) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (ESaracay) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (ESaracay) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (ESaracay) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (ESaracay) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done.

**Review by tylerhou — APPROVED** [review]

**Pull request merged**